File: app/controllers/iop/collaboration/impl/CollaborativeExternalCalendarController.java
package controllers.iop.collaboration.impl;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.typesafe.config.ConfigFactory;
import exceptions.NotFoundException;
import impl.CalendarHandler;
import io.ebean.Ebean;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Collection;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import java.util.stream.Collector;
import javax.inject.Inject;
import models.Exam;
import models.ExamEnrolment;
import models.User;
import models.json.CollaborativeExam;
import org.joda.time.DateTime;
import org.joda.time.format.ISODateTimeFormat;
import play.libs.Json;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;
import sanitizers.Attrs;
import sanitizers.ExternalCalendarReservationSanitizer;
import security.Authenticated;
import util.config.ConfigReader;
import util.datetime.DateTimeUtils;

public class CollaborativeExternalCalendarController extends CollaborativeCalendarController {

    @Inject
    CalendarHandler calendarHandler;

    @Inject
    private ConfigReader configReader;

    @Authenticated
    @With(ExternalCalendarReservationSanitizer.class)
    @Restrict({ @Group("STUDENT") })
    public CompletionStage<Result> requestReservation(Http.Request request) {
        if (!configReader.isVisitingExaminationSupported()) {
            return wrapAsPromise(forbidden("Feature not enabled in the installation"));
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        // Parse request body
        String orgRef = request.attrs().get(Attrs.ORG_REF);
        String roomRef = request.attrs().get(Attrs.ROOM_REF);
        DateTime start = request.attrs().get(Attrs.START_DATE);
        DateTime end = request.attrs().get(Attrs.END_DATE);
        Long examId = request.attrs().get(Attrs.EXAM_ID);
        Collection<Long> sectionIds = request.attrs().get(Attrs.SECTION_IDS);
        DateTime now = DateTimeUtils.adjustDST(DateTime.now());

        CollaborativeExam ce = Ebean.find(CollaborativeExam.class, examId);
        if (ce == null) {
            return wrapAsPromise(notFound("sitnet_error_exam_not_found"));
        }
        final ExamEnrolment enrolment = Ebean
            .find(ExamEnrolment.class)
            .fetch("reservation")
            .where()
            .eq("user.id", user.getId())
            .eq("collaborativeExam.id", examId)
            .disjunction()
            .isNull("reservation")
            .gt("reservation.startAt", now.toDate())
            .endJunction()
            .findOne();
        if (enrolment == null) {
            return wrapAsPromise(notFound("sitnet_error_exam_not_found"));
        }
        return downloadExam(ce)
            .thenComposeAsync(
                result -> {
                    if (result.isEmpty()) {
                        return wrapAsPromise(notFound("sitnet_error_exam_not_found"));
                    }
                    Exam exam = result.get();
                    Optional<Result> badEnrolment = checkEnrolment(enrolment, exam, user);
                    if (badEnrolment.isPresent()) {
                        return wrapAsPromise(badEnrolment.get());
                    }
                    // Make ext request here
                    // Lets do this
                    URL url;
                    try {
                        url = parseUrl(orgRef, roomRef);
                    } catch (MalformedURLException e) {
                        throw new RuntimeException(e);
                    }
                    String homeOrgRef = ConfigFactory.load().getString("sitnet.integration.iop.organisationRef");
                    ObjectNode body = Json.newObject();
                    body.put("requestingOrg", homeOrgRef);
                    body.put("start", ISODateTimeFormat.dateTime().print(start));
                    body.put("end", ISODateTimeFormat.dateTime().print(end));
                    body.put("user", user.getEppn());
                    body.set(
                        "optionalSections",
                        sectionIds.stream().collect(Collector.of(Json::newArray, ArrayNode::add, ArrayNode::add))
                    );

                    WSRequest wsRequest = wsClient.url(url.toString());
                    return wsRequest
                        .post(body)
                        .thenComposeAsync(
                            response -> {
                                JsonNode root = response.asJson();
                                if (response.getStatus() != Http.Status.CREATED) {
                                    return wrapAsPromise(
                                        internalServerError(root.get("message").asText("Connection refused"))
                                    );
                                }
                                return calendarHandler
                                    .handleExternalReservation(
                                        enrolment,
                                        exam,
                                        root,
                                        start,
                                        end,
                                        user,
                                        orgRef,
                                        roomRef,
                                        sectionIds
                                    )
                                    .thenApplyAsync(
                                        err -> {
                                            if (err.isEmpty()) {
                                                return created(root.get("id"));
                                            }
                                            return internalServerError();
                                        }
                                    );
                            }
                        );
                }
            );
    }

    @Authenticated
    @Restrict({ @Group("STUDENT") })
    public CompletionStage<Result> requestSlots(
        Long examId,
        String roomRef,
        Optional<String> org,
        Optional<String> date,
        Http.Request request
    ) {
        if (org.isPresent() && date.isPresent()) {
            User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
            CollaborativeExam ce = Ebean.find(CollaborativeExam.class, examId);
            if (ce == null) {
                return wrapAsPromise(notFound("sitnet_error_exam_not_found"));
            }
            ExamEnrolment enrolment = getEnrolledExam(examId, user);
            if (enrolment == null) {
                return wrapAsPromise(forbidden("sitnet_error_enrolment_not_found"));
            }
            return downloadExam(ce)
                .thenComposeAsync(
                    result -> {
                        if (result.isEmpty()) {
                            return wrapAsPromise(notFound("sitnet_error_exam_not_found"));
                        }
                        Exam exam = result.get();
                        if (!exam.hasState(Exam.State.PUBLISHED)) {
                            return wrapAsPromise(notFound("sitnet_error_exam_not_found"));
                        }
                        // Also sanity check the provided search date
                        try {
                            calendarHandler.parseSearchDate(date.get(), exam, null);
                        } catch (NotFoundException e) {
                            return wrapAsPromise(notFound());
                        }
                        // Ready to shoot
                        String start = ISODateTimeFormat.dateTime().print(new DateTime(exam.getExamActiveStartDate()));
                        String end = ISODateTimeFormat.dateTime().print(new DateTime(exam.getExamActiveEndDate()));
                        Integer duration = exam.getDuration();
                        URL url = parseUrl(org.get(), roomRef, date.get(), start, end, duration);
                        WSRequest wsRequest = wsClient
                            .url(url.toString().split("\\?")[0])
                            .setQueryString(url.getQuery());
                        Function<WSResponse, Result> onSuccess = response -> {
                            JsonNode root = response.asJson();
                            if (response.getStatus() != Http.Status.OK) {
                                return internalServerError(root.get("message").asText("Connection refused"));
                            }
                            Set<CalendarHandler.TimeSlot> slots = calendarHandler.postProcessSlots(
                                root,
                                date.get(),
                                exam,
                                user
                            );
                            return ok(Json.toJson(slots));
                        };
                        return wsRequest.get().thenApplyAsync(onSuccess);
                    }
                );
        }
        return wrapAsPromise(badRequest());
    }

    private ExamEnrolment getEnrolledExam(Long examId, User user) {
        DateTime now = DateTimeUtils.adjustDST(DateTime.now());
        return Ebean
            .find(ExamEnrolment.class)
            .where()
            .eq("user", user)
            .eq("collaborativeExam.id", examId)
            .disjunction()
            .isNull("reservation")
            .gt("reservation.startAt", now.toDate())
            .endJunction()
            .findOne();
    }

    private static URL parseUrl(String orgRef, String facilityRef) throws MalformedURLException {
        return new URL(
            ConfigFactory.load().getString("sitnet.integration.iop.host") +
            String.format("/api/organisations/%s/facilities/%s/reservations", orgRef, facilityRef)
        );
    }

    private static URL parseUrl(
        String orgRef,
        String facilityRef,
        String date,
        String start,
        String end,
        int duration
    ) {
        String url =
            ConfigFactory.load().getString("sitnet.integration.iop.host") +
            String.format("/api/organisations/%s/facilities/%s/slots", orgRef, facilityRef) +
            String.format("?date=%s&startAt=%s&endAt=%s&duration=%d", date, start, end, duration);
        try {
            return new URL(url);
        } catch (MalformedURLException e) {
            throw new RuntimeException(e);
        }
    }
}

File: app/controllers/iop/collaboration/impl/CollaborativeExternalCalendarController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.collaboration.impl

import impl.CalendarHandler
import io.ebean.DB
import miscellaneous.config.ConfigReader
import miscellaneous.datetime.DateTimeHandler
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.enrolment.ExamEnrolment
import models.exam.Exam
import models.iop.CollaborativeExam
import models.user.{Role, User}
import org.joda.time.DateTime
import org.joda.time.format.ISODateTimeFormat
import play.api.Logging
import play.api.libs.json.{JsArray, JsValue, Json, Writes}
import play.api.libs.ws.{JsonBodyWritables, WSClient}
import play.api.mvc.*
import security.scala.Auth.{AuthenticatedAction, authorized}
import security.scala.{Auth, AuthExecutionContext}
import validation.scala.calendar.{ExternalReservationDTO, ReservationCreationFilter}
import validation.scala.core.ScalaAttrs

import java.net.{MalformedURLException, URI, URL}
import javax.inject.Inject
import scala.concurrent.Future
import scala.jdk.CollectionConverters.*
import scala.jdk.FutureConverters.*

class CollaborativeExternalCalendarController @Inject() (
    authenticated: AuthenticatedAction,
    calendarHandler: CalendarHandler,
    wsClient: WSClient,
    configReader: ConfigReader,
    dateTimeHandler: DateTimeHandler,
    enrolmentHandler: miscellaneous.enrolment.EnrolmentHandler,
    collaborationController: CollaborationController, // For downloadExam
    val controllerComponents: ControllerComponents,
    implicit val ec: AuthExecutionContext
) extends BaseController
    with DbApiHelper
    with JavaApiHelper
    with JsonBodyWritables
    with Logging:

  def requestReservation(): Action[JsValue] =
    authenticated(parse.json)
      .andThen(authorized(Seq(Role.Name.STUDENT)))
      .andThen(ReservationCreationFilter.forExternal())
      .async { request =>
        if !configReader.isVisitingExaminationSupported then
          Future.successful(Forbidden("Feature not enabled in the installation"))
        else
          val user = request.attrs(Auth.ATTR_USER)

          // Parse request body
          val ExternalReservationDTO(orgRef, roomRef, examId, start, end, sectionIds) =
            request.attrs(ScalaAttrs.ATTR_EXT_RESERVATION)
          val now = dateTimeHandler.adjustDST(DateTime.now())

          val ceOpt = Option(DB.find(classOf[CollaborativeExam], examId))
          ceOpt match
            case None => Future.successful(NotFound("i18n_error_exam_not_found"))
            case Some(ce) =>
              val enrolmentOpt = Option(
                DB.find(classOf[ExamEnrolment])
                  .fetch("reservation")
                  .where()
                  .eq("user.id", user.getId)
                  .eq("collaborativeExam.id", examId)
                  .disjunction()
                  .isNull("reservation")
                  .gt("reservation.startAt", now.toDate)
                  .endJunction()
                  .findOne()
              )

              enrolmentOpt match
                case None => Future.successful(NotFound("i18n_error_exam_not_found"))
                case Some(enrolment) =>
                  collaborationController.downloadExam(ce).asScala.flatMap { examOpt =>
                    if examOpt.isEmpty then Future.successful(NotFound("i18n_error_exam_not_found"))
                    else
                      val exam = examOpt.get
                      // Check enrolment validity
                      val oldReservation = enrolment.getReservation
                      val checkResult =
                        if exam.getState == Exam.State.STUDENT_STARTED ||
                          (oldReservation != null && oldReservation.toInterval.isBefore(DateTime.now()))
                        then Some(Forbidden("i18n_reservation_in_effect"))
                        else if oldReservation == null && !enrolmentHandler.isAllowedToParticipate(exam, user) then
                          Some(Forbidden("i18n_no_trials_left"))
                        else None

                      checkResult match
                        case Some(errorResult) => Future.successful(errorResult)
                        case None              =>
                          // Make ext request here
                          val url =
                            try parseUrl(orgRef, roomRef)
                            catch case e: MalformedURLException => throw new RuntimeException(e)

                          val homeOrgRef = configReader.getHomeOrganisationRef
                          val body = Json.obj(
                            "requestingOrg"    -> homeOrgRef,
                            "start"            -> ISODateTimeFormat.dateTime().print(start),
                            "end"              -> ISODateTimeFormat.dateTime().print(end),
                            "user"             -> user.getEppn,
                            "optionalSections" -> JsArray(sectionIds.getOrElse(List.empty).map(id => Json.toJson(id)))
                          )

                          wsClient
                            .url(url.toString)
                            .post(body)
                            .flatMap { response =>
                              val root = response.json
                              if response.status != CREATED then
                                Future.successful(
                                  InternalServerError((root \ "message").asOpt[String].getOrElse("Connection refused"))
                                )
                              else
                                calendarHandler
                                  .handleExternalReservation(
                                    enrolment,
                                    exam,
                                    root,
                                    start,
                                    end,
                                    user,
                                    orgRef,
                                    roomRef,
                                    sectionIds.getOrElse(List.empty)
                                  )
                                  .map { err =>
                                    if err.isEmpty then Created((root \ "id").as[JsValue])
                                    else InternalServerError("Failed to handle external reservation")
                                  }
                            }
                  }
      }

  def requestSlots(
      examId: Long,
      roomRef: String,
      org: Option[String],
      date: Option[String]
  ): Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.STUDENT))).async { request =>
      (org, date) match
        case (Some(orgValue), Some(dateValue)) =>
          val user  = request.attrs(Auth.ATTR_USER)
          val ceOpt = Option(DB.find(classOf[CollaborativeExam], examId))

          ceOpt match
            case None => Future.successful(NotFound("i18n_error_exam_not_found"))
            case Some(ce) =>
              val enrolmentOpt = Option(getEnrolledExam(examId, user))

              enrolmentOpt match
                case None => Future.successful(Forbidden("i18n_error_enrolment_not_found"))
                case Some(_) =>
                  collaborationController.downloadExam(ce).asScala.flatMap { examOpt =>
                    if examOpt.isEmpty then Future.successful(NotFound("i18n_error_exam_not_found"))
                    else
                      val exam = examOpt.get
                      if !exam.hasState(Exam.State.PUBLISHED) then
                        Future.successful(NotFound("i18n_error_exam_not_found"))
                      else
                        // Also, sanity-check the provided search date
                        try
                          calendarHandler.parseSearchDate(dateValue, exam, null)

                          // Ready to shoot
                          val start    = ISODateTimeFormat.dateTime().print(new DateTime(exam.getPeriodStart))
                          val end      = ISODateTimeFormat.dateTime().print(new DateTime(exam.getPeriodEnd))
                          val duration = exam.getDuration
                          val url      = parseUrl(orgValue, roomRef, dateValue, start, end, duration)

                          val wsRequest = wsClient
                            .url(url.toString.split("\\?")(0))
                            .withQueryStringParameters(
                              url.getQuery.split("&").map { param =>
                                val parts = param.split("=")
                                parts(0) -> parts(1)
                              }*
                            )

                          wsRequest.get().map { response =>
                            val root = response.json
                            if response.status != OK then
                              InternalServerError((root \ "message").asOpt[String].getOrElse("Connection refused"))
                            else
                              val slots = calendarHandler.postProcessSlots(root, dateValue, exam, user)
                              Ok(Json.toJson(slots.toSeq))
                          }
                        catch
                          case _: IllegalArgumentException =>
                            Future.successful(NotFound("Invalid date"))
                  }
        case _ => Future.successful(BadRequest("Missing required parameters"))
    }

  private def getEnrolledExam(examId: Long, user: User): ExamEnrolment =
    val now = dateTimeHandler.adjustDST(DateTime.now())
    DB.find(classOf[ExamEnrolment])
      .where()
      .eq("user", user)
      .eq("collaborativeExam.id", examId)
      .disjunction()
      .isNull("reservation")
      .gt("reservation.startAt", now.toDate)
      .endJunction()
      .findOne()

  @throws[MalformedURLException]
  private def parseUrl(orgRef: String, facilityRef: String): URL =
    URI
      .create(
        configReader.getIopHost +
          f"/api/organisations/$orgRef/facilities/$facilityRef/reservations"
      )
      .toURL

  private def parseUrl(
      orgRef: String,
      facilityRef: String,
      date: String,
      start: String,
      end: String,
      duration: Int
  ): URL =
    val url =
      configReader.getIopHost +
        f"/api/organisations/$orgRef/facilities/$facilityRef/slots" +
        f"?date=$date&startAt=$start&endAt=$end&duration=$duration"
    try URI.create(url).toURL
    catch case e: MalformedURLException => throw new RuntimeException(e)
